File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
//null-1-2-3
// 1-null -
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
       ListNode* prev = nullptr;
       ListNode* cur = head;

       while (cur != nullptr) {
        ListNode* temp = cur-> next; //1. save original next
        cur->next = prev; // 2. reverse the pointer
        prev = cur; // more prev forward where cur is
        cur = temp; // move cur forwrad
       }
       return prev;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # get the middle, then starting tht mode reverse list then to new list add one from first list one from reversed
        # [0, 1, 2, 3s,   4, 5, 6f]. 
       
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        cur = slow.next
        slow.next = None
        # reverse second half
        prev = None
        while cur:
            next_node = cur.next
            cur.next = prev
            prev = cur
            cur = next_node

        first = head   
        second = prev 
        while second:
            f_next, s_next = first.next, second.next
            first.next = second
            second.next = f_next
            first, second =  f_next, s_next

